transactions/WFLY-17742_SupportTransactionsDuringGracefulShutdown

```diff
@@ -0,0 +1,155 @@
+
+_1.2 There are additional complications when hosting entities crash (i.e. an unintentional shutdown occurs). This extreme scenario is beyond what WildFly can handle and the hosting cloud environment must provide the ability to recover the hosting entity from crashes._
+
+The state of the hosting entity (e.g. its file system, ip address, etc.) should be recovered by the cloud environment without the intervention of the user. In K8s, for example, the employment of StatefulSet creates a stable environment where crashes are handled automatically. If a particular cloud environment does not provide a guarantee similar to StatefulSet, it is not possible to employ WildFly’s Transaction Subsystem.
+
+*As a solution to solve problem 1.2 cannot be developed in WildFly (as it is impossible to control the lifecycle of the host from WildFly), this proposal is only concerned with problem 1.1.*
+
+== Issue Metadata
+
+=== Issue
+
+* [WFLY-17742](https://issues.jboss.org/browse/WFLY-17742)
+
+=== Dev Contacts
+
+* mailto:{jfinelli@redhat.com}[{Manuel Finelli}]
+
+=== QE Contacts
+
+* mailto:{istraka@redhat.com}[{Ivan Straka}]
+
+=== Testing By
+
+* [ ] Engineering
+* [X] QE
+
+=== Affected Projects or Components
+
+* [wildfly-core](https://github.com/wildfly/wildfly-core)
+* [wildfly](https://github.com/wildfly/wildfly)
+* [narayana](https://github.com/jbosstm/narayana)
+
+=== Relevant Installation Types
+* [x] Traditional standalone server (unzipped or provisioned by Galleon)
+
+* [x] Managed domain
+
+* [x] OpenShift s2i
+
+* [x] Bootable jar
+
+== Requirements
+
+=== Hard Requirements
+
+* The Transactions subsystem should become aware of how many in-doubt and heuristic transactions there are in the Object Store and delay its suspension as long as all transactions are completed. This would guarantee data integrity in all scenarios, including situations when a negative timeout (i.e. indefinite waiting) is employed (e.g. in cloud environments)
+* With regards to the suspension logic of the EJB subsystem ([reference](https://github.com/wildfly/wildfly/blob/main/ejb3/src/main/java/org/jboss/as/ejb3/suspend/EJBSuspendHandlerService.java#L203)), the suspension handler of this subsystem should retain control (i.e. it should not return) as long as there are in-doubt transactions and heuristic transactions to be completed. This would guarantee data integrity, especially in situations when a negative timeout (i.e. indefinite waiting) is employed (e.g. in cloud environments)
+* Integrating with the existing logging mechanism of WildFly, users should be notified when subsystems are delaying WildFly’s suspension. This is especially true when a negative timeout is used (i.e. indefinite graceful shutdown). Moreover, internal details like 'ServerActivity' do not need to be prominent. Part of working out such a notification would be how to consistently describe a ServerActivity to a user in an understandable manner without exposing internal technical implementation details.
```

Comment:
I interpreted the word "users" in the sentence "users should be notified ..." as referring to a WildFly subsystem. Although the sentence was not clear I assumed that end users and applications have no need to be notified, hence I guessed `user == subsystem`. If end users or applications do need to be notified then the document needs to indicate why.

Reply:
I said it's not clear what the actual requirement is and I don't know what your intention was so I cannot suggest how it could be reworded.

Reply:
I understand this requirement to be usability related. If a subsystem is delaying WildFly's suspension it needs to be logged but the part that isn't clear is why it needs to be logged and and who "users" include. I could imagine "users" could specifically be the "admins of a WildFly cluster" (and if so, that wording might replace "users"), and the motiviation to do that would be usability to not just hang without informing the admin why it is hanging.shutdown.

It may be possible to consider simply removing the part `Moreover, internal details like 'ServerActivity' do not need to be prominent. Part of working out such a notification would be how to consistently describe a ServerActivity to a user in an understandable manner without exposing internal technical implementation details.`

Reply:
Thanks for your message Tom :-)

IIRC, I added this part:
```
Moreover, internal details like 'ServerActivity' do not need to be prominent. Part of working out such a notification would be how to consistently describe a ServerActivity to a user in an understandable manner without exposing internal technical implementation details.
```
to address Brian's comment (which was posted in the related JIRA):
```
Note that to the extent that some of this ends up being end-user facing (e.g. notifying users of what is delaying things), then there we don't want internal details like 'ServerActivity' being prominent. Part of working out such a notification would need to be working out how to consistently describe a ServerActivity to a user in an understandable manner.
```
Maybe, as you suggested, replacing "users" with "admins of a WildFly cluster" could help clarify this part of the proposal? WDYT @tomjenkinson, @mmusgrov?

Reply:
Thank you for citing that reference.

Do you know if there is a standard mentioned anywhere with that ServerActivity type of thing that developers should already be aware of as it may not be necessary to add it to this specific discussion but just something to be aware of? And even if not, perhaps that could be added as a general guidance outside of this proposal, like in some kind of WildFly Logging Standards for developers documentation?

Is this feature intended for usage in a non-cloud environment? If so, I would try to have the requirement just be:

`Integrating with the existing logging mechanism of WildFly is required in order to notify users when something is delaying WildFly’s suspension.`

but if it's just for the cloud case (I guess not though) then `users` could be made more specificalised.

Reply:
> Thank you for citing that reference.
> 
> Do you know if there is a standard mentioned anywhere with that ServerActivity type of thing that developers should already be aware of as it may not be necessary to add it to this specific discussion but just something to be aware of? And even if not, perhaps that could be added as a general guidance outside of this proposal, like in some kind of WildFly Logging Standards for developers documentation?

As far as I know, I don't think there is any mention about ServerActivity in WildFly's documentation. If I have to guess, it's a technical detail that is too deep in the code to document it. Having said that, I don't think it's a problem removing this specific part from the proposal.

> Is this feature intended for usage in a non-cloud environment? If so, I would try to have the requirement just be:
> 
> `Integrating with the existing logging mechanism of WildFly is required in order to notify users when something is delaying WildFly’s suspension.`
> 
> but if it's just for the cloud case (I guess not though) then `users` could be made more specificalised.

This proposal covers both cloud and non-cloud environments thus we can use `Integrating with the existing logging mechanism of WildFly is required in order to notify users when something is delaying WildFly’s suspension.`

Reply:
> 

> This proposal covers both cloud and non-cloud environments thus we can use `Integrating with the existing logging mechanism of WildFly is required in order to notify users when something is delaying WildFly’s suspension.`

Thanks @jmfinelli - I think that requirement would be clear. Do you agree @mmusgrov ?

Reply:
I thought that was already in the proposal. But if storing the transaction logs is not part of the solution space then I guess integrating with the existing logging mechanism is one solution to the requirement.

Reply:
> Maybe, as you suggested, replacing "users" with "admins of a WildFly cluster" could help clarify this part of the proposal? WDYT @tomjenkinson, @mmusgrov?

If that's what a "user" is in this context then +1